feat: Add model label for vllm backend metrics

tzulingk opened the pull request:
#### Overview:

Implement add_labels for Endpoint and add model name to StatLoggerFactory and DynamoStatLoggerPublisher.

#### Details:

Simply adding the model labels to the Endpoint doesn’t make the metrics include the model label.
This happens because the code registers the same metric multiple times. The first registration doesn’t have the model label, and by the time we add the model label to the Endpoint through add_labels, it’s already too late.

The first time is added through 
```
/// components/backends/vllm/src/dynamo/vllm/main.py init()
factory = StatLoggerFactory(component, config.engine_args.data_parallel_rank or 0)
```
The second time is added through 
```
generate_endpoint.serve_endpoint(handler.generate, graceful_shutdown=False),
clear_endpoint.serve_endpoint(handler.clear_kv_blocks),
```
So we need to pass the model name into StatLoggerFactory and DynamoStatLoggerPublisher

#### Where should the reviewer start?

components/backends/vllm/src/dynamo/vllm/publisher.py: Add model name to DynamoStatLoggerPublisher and StatLoggerFactory
lib/runtime/src/component.rs: Implement add_labels for Endpoint.

#### Related Issues: (use one of the action keywords Closes / Fixes / Resolves / Relates to)

DIS-444 The model label is missing from the backend metrics


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

#### Example

```
# Start frontend
python -m dynamo.frontend &

# Enable per-component system/metrics http servers
export DYN_SYSTEM_ENABLED=true

# Start vllm worker 1, with metrics enabled, and unique port for metrics server
DYN_SYSTEM_PORT=9090 CUDA_VISIBLE_DEVICES=0 python -m dynamo.vllm --model Qwen/Qwen3-0.6B --enforce-eager --connector none &

# Start vllm worker 2, with metrics enabled, and unique port for metrics server
DYN_SYSTEM_PORT=9091 CUDA_VISIBLE_DEVICES=1 python -m dynamo.vllm --model Qwen/Qwen3-0.6B --enforce-eager --connector none &

# Check metrics outputs for each worker
curl -s localhost:9090/metrics | grep dynamo_component_requests_total | grep generate | grep model
# dynamo_component_requests_total{dynamo_component="backend",dynamo_endpoint="generate",dynamo_namespace="dynamo",model="Qwen/Qwen3-0.6B"} 0

curl -s localhost:9091/metrics | grep dynamo_component_requests_total | grep generate | grep model
# dynamo_component_requests_total{dynamo_component="backend",dynamo_endpoint="generate",dynamo_namespace="dynamo",model="Qwen/Qwen3-0.6B"} 0

# Send inference request
curl -X POST http://localhost:8080/v1/chat/completions   -H 'Content-Type: application/
json'   -d '{
    "model": "Qwen/Qwen3-0.6B",
    "messages": [
      { "role": "user", "content": "Tell me a story about a brave cat" }
    ],
    "stream": false,
    "max_tokens": 1028
  }'

# Check metrics outputs to see that one worker got updated, and the other did not
curl -s localhost:9090/metrics | grep dynamo_component_requests_total | grep generate | grep model
# dynamo_component_requests_total{dynamo_component="backend",dynamo_endpoint="generate",dynamo_namespace="dynamo",model="Qwen/Qwen3-0.6B"} 0
(venv) root@4u4g-0073:/workspace# curl -s localhost:9091/metrics | grep dynamo_component_requests_total | grep generate | grep model
# dynamo_component_requests_total{dynamo_component="backend",dynamo_endpoint="generate",dynamo_namespace="dynamo",model="Qwen/Qwen3-0.6B"} 1

# Success: one of the workers received the request and incremented the metric
# With a 2nd request (assuming round robin routing), you will see the other worker's metric increment this time.  
```

## Summary by CodeRabbit

- New Features
  - Endpoints can now be labeled (e.g., model) for per-model routing and metrics across runtime, metrics, and vLLM.
  - Python: added Endpoint.add_labels(labels) to attach constant labels.
  - Example clients/servers updated to annotate “generate” endpoints with model labels.

- Tests
  - Added unit tests verifying label injection and propagation into Prometheus metrics.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

tzulingk committed:
Implement add_labels for Endpoint and add model name to StatLoggerFactory and DynamoStatLoggerPublisher

tzulingk committed:
format the codes

tzulingk committed:
remove unused import

nnshah1 commented on lib/runtime/examples/service_metrics/src/bin/service_client.rs:
```diff
@@ -33,7 +33,11 @@ async fn app(runtime: Runtime) -> Result<()> {
-    let client = component.endpoint("generate").client().await?;
+    let client = component
+        .endpoint("generate")
+        .add_labels(&[("model", "service_metrics_model")])?
```
same as above

nnshah1 reviewed (changes requested):
> This happens because the code registers the same metric multiple times. The first registration doesn’t have the model label, and by the time we add the model label to the Endpoint through add_labels, it’s already too late.

This point is a little confusing - IIUC - what this means is that all endpoints served by a component require the same set of labels? We can note this for now - but I can imagine this being error prone - maybe we could add an error in case a metric is registered with a different set of labels?

The other part that can be confusing here is adding labels to endpoint clients - as we don't collect any metrics there - 

With that in mind - and since we are refactoring here in any case - what if we added labels as an optional field to "start()"

https://github.com/ai-dynamo/dynamo/blob/fb10ffb55213e24966e28e13850cc3e7964dc8ab/lib/runtime/src/component/endpoint.rs#L61

Essentially this would be: (pseudo code):

```
start(runtime_metrics_labels=[(model='foo')])
```

and this is where metrics are added via "add_metrics" to the push_handler - ( I believe here https://github.com/ai-dynamo/dynamo/blob/fb10ffb55213e24966e28e13850cc3e7964dc8ab/lib/runtime/src/pipeline/network.rs#L303)

and here:

https://github.com/ai-dynamo/dynamo/blob/fb10ffb55213e24966e28e13850cc3e7964dc8ab/lib/runtime/src/component/endpoint.rs#L78

Note: This is from a cursory look - and this may not be possible - so if it doesn't make sense to add them in this lazy fashion (i.e. where the start()) is - this isn't blocking

The benefit here though is that it would avoid any possibility of someone adding labels to a endpoint before it is started - so we don't have to worry about someone mistakenly adding labels to a client and getting confused.

keivenchang commented on lib/runtime/examples/hello_world/src/bin/client.rs:
```diff
@@ -32,6 +32,7 @@ async fn app(runtime: Runtime) -> Result<()> {
         .namespace(DEFAULT_NAMESPACE)?
         .component("backend")?
         .endpoint("generate")
+        .add_labels(&[("model", "hello_world_model")])?
```
* This is an example program:
  - it doesn't get run on CI tests
  - it doesn't have anything to do with models. 
  - therefore, it's not providing any value except for validating that add_labels works, when tested _manually_
  - I would remove it from all example code

Having that said, SUPPOSE this example serves a model. Then, the entire endpoint called "generate" is serving a model, right?

nnshah1 replied:
> Also, here, the client is not calling service_builder and endpoint_builder so it's not going to have a Prometheus metrics.

I think this is my point - adding labels to something that doesn't have metrics is a little confusing - it's not that it is a an example - its that the server portion is the one that shows metrics and the client doesn't - so the api can lead to confusion here -

keivenchang commented on lib/bindings/python/rust/llm/kv.rs:
```diff
@@ -55,17 +55,18 @@ impl WorkerMetricsPublisher {
         &self,
         py: Python<'p>,
         component: Component,
+        model: Option<String>,
```
Hold on. I thought we want to detach the concept of a model, away from the concept of the Runtime Hierarchy. And to do that, use labels (or add_labels).

nnshah1 replied:
this is ok either way I believe - once we are in the "kv publisher create endpoint" - then we are creating one for a model - the place in the code here already assume llm model as it creates a publisher for kv metrics

keivenchang commented on lib/runtime/examples/hello_world/src/bin/server.rs:
```diff
@@ -73,6 +73,7 @@ async fn backend(runtime: DistributedRuntime) -> Result<()> {
         .create()
         .await?
         .endpoint("generate")
+        .add_labels(&[("model", "hello_world_model")])?
```
This example doesn't serve a model. It's going to confuse new programmers who think it does.

nnshah1 replied:
I think this is ok either way - this could show generally how to add labels - could be "echo" - as this service echos .... but ok to remove as well.

keivenchang committed:
refactor: change create_endpoint from model to labels parameter

- Change parameter from model to labels for better semantic clarity
- Removed model concept from example files
- Maintain same API contract for Python users

keivenchang committed:
refactor: update metrics endpoint creation to use labels instead of model parameter

keivenchang committed:
style: fix formatting in service_metrics example

tzulingk committed:
Add metrics labels at EndpointConfigBuilder::start()

grahamking commented:
Could we back up a bit and think about labels? If feels like you're gradually adding labels as a parameter to every function in Dynamo.

Each piece of data should have a single owner, and a single way to create it. That makes the system much easier to understand. Figure out why a user would want a label - is "model" the only one? - then figure out which piece of Dynamo owns that.

Being as labels are a metrics thing, they should most likely be owned by the metrics registry or similar top level metrics object. Maybe give that a static method, then from anywhere you can do `MetricsRegistry::add_label(name, value)`.

tzulingk committed:
Add metric_labels in EndpointConfig.

tzulingk committed:
Merge remote-tracking branch 'origin/main' into tzulingk/model_label_dis_444

nnshah1 commented:
> Can you call them `metric_labels` in most places (so not necessary in `add_metrics` for example)? We may have other kinds of labels (e.g. service level, K8s things).
> 
> Can you also remove the `add_labels` from namespace / component / endpoint? Those won't be used now that we pass directly.
> 
> Finally it's a pity we have to motify `serve_endpoint` but I couldn't see an obvious way around it. @nnshah1 Do you remember how we looked at it on the call?

I think unfortunately serve_endpoint (python side) is where we construct the builder and don't see a way around it in the short term, serve_endpoint is kind of of builder + start all in one. 

I think as long as its optional we can change on the python side in the future -

tzulingk commented:
> Can you call them `metric_labels` in most places (so not necessary in `add_metrics` for example)? We may have other kinds of labels (e.g. service level, K8s things).
> 
> Can you also remove the `add_labels` from namespace / component / endpoint? Those won't be used now that we pass directly.
> 
> Finally it's a pity we have to motify `serve_endpoint` but I couldn't see an obvious way around it. @nnshah1 Do you remember how we looked at it on the call?

I believe add_labels are already been removed. Could you point me to where you see it?
I will change the labels to metric_labels

tzulingk committed:
Change labels to metrics_labels

tzulingk committed:
Merge branch 'main' into tzulingk/model_label_dis_444

rmccorm4 commented:
@keivenchang please re-review or remove "requested changes"

tzulingk committed:
Cargo fmt

tzulingk committed:
Merge branch 'main' into tzulingk/model_label_dis_444

tzulingk committed:
Merge branch 'main' into tzulingk/model_label_dis_444

tzulingk committed:
Use different endpoint name to aavoid the duplicated metrics registrations.

tzulingk committed:
Merge branch 'main' into tzulingk/model_label_dis_444

rmccorm4 commented on tests/router/test_router_e2e_with_mockers.py:
```diff
@@ -216,7 +216,7 @@ def test_mocker_kv_router(request, runtime_services):
         for i in range(NUM_MOCKERS):
             # Use unique endpoints for each mocker
-            endpoint = "dyn://test-namespace.mocker.generate"
+            endpoint = f"dyn://test-namespace.mocker-{i}.generate"
```
I think something is wrong here if starting two models with the same endpoint is causing an issue. This should be well supported behavior in Dynamo - having multiple replicas of a model behind same dynamo endpoint is how we route (random, round robin, kv-aware, etc.). If all endpoints need to be uniquely named, then we break routing.

tzulingk replied:
### Find out how duplicated registrations happen from the 2 paths of test codes below
```
 mocker = MockerProcess(request, endpoint, mocker_args_file)
 mocker.__enter__()
```
- First Registration
ManagedProcess._\__enter__\_()
-> ManagedProcess._start_process()
-> self.proc = subprocess.Popen(self.command, ...) // python -m dynamo.mocker --model-path ... --endpoint ...
-> worker() // components/backends/mocker/src/dynamo/mocker/main.py
-> run_input() // dynamo_llm::entrypoint::input::run_input()
-> endpoint::run()
-> endpoint.endpoint_builder().handler(ingress_chat).start(); // register metrics

- Second Registration
ManagedProcess._\__enter__\_()
-> ManagedProcess._start_process()
-> self.proc = subprocess.Popen(self.command, ...) // python -m dynamo.mocker --model-path ... --endpoint ...
-> worker() // components/backends/mocker/src/dynamo/mocker/main.py
-> engine_config = await make_engine(runtime, entrypoint_args) // bind to make_engine in lib/bindings/python/rust/llm/entrypoint.rs
-> select_engine()
-> dynamo_llm::mocker::engine::make_mocker_engine()
-> MockVllmEngine::new()
-> AnnotatedMockEngine::new()
-> MockVllmEngine::start() 
-> MockVllmEngine::start_metrics_publishing() 
```
let metrics_labels = Some(&[("model", "mock_model")][..]);
if let Err(e) = publisher
    .create_endpoint(comp.clone(), metrics_labels) // register metrics internally
    .await
```

So by removing the metrics_labels in MockVllmEngine, we can avoid having different labels set for the same metric.

rmccorm4 reviewed (changes requested):
Requesting changes based on this recent commit/thread: https://github.com/ai-dynamo/dynamo/pull/2474#discussion_r2289733730

Hopefully an easy fix in the morning - but I don't think fix here should be to make endpoint names unique. We need to support multiple component replicas having the same endpoint name.

tzulingk committed:
Remove the metrics_labels in MockVllmEngine, so that we will have the same description for the same metrics

tzulingk committed:
format tests/router/test_router_e2e_with_mockers.py

tzulingk committed:
Merge branch 'main' into tzulingk/model_label_dis_444

rmccorm4 commented on components/backends/vllm/src/dynamo/vllm/main.py:
```diff
@@ -150,8 +150,14 @@ async def init_prefill(runtime: DistributedRuntime, config: Config):
+            generate_endpoint.serve_endpoint(
+                handler.generate,
+                graceful_shutdown=True,
+                metrics_labels=[("model", config.model)],
```
Can you confirm that you can start 2 vllm workers with the same model without any errors?

tzulingk replied:
Yes, I confirm that there's no error. And I queried the frontend with 
```
curl -s localhost:8080/v1/chat/completions -H "Content-Type: application/json" -d '{
  "model": "Qwen/Qwen3-0.6B",
  "messages": [
    {
      "role": "developer",
      "content": "You are a helpful assistant."
    },
    {
      "role": "user",
      "content": "Hello!"
    }
  ],
  "stream": false,
  "max_tokens": 300
}'
```
And get result
```
{"id":"chatcmpl-e398eff1-6a15-457c-9db9-6f5385284d03","choices":[{"index":0,"message":{"content":"<think>\nOkay, the user just said \"Hello!\" so I need to respond appropriately. Let me think about the best way to greet them. Since they started with a greeting, a simple and friendly reply would be good. Maybe something like \"Hi there!\" to keep it light and welcoming. I should make sure the tone is positive and open-ended to encourage any further conversation. Let me check if there's anything else they might need, but since they just said hello, maybe just a standard reply is sufficient. Alright, I'll go with \"Hi there!\" as the response.\n</think>\n\nHi there! 😊 How can I assist you today?","role":"assistant","reasoning_content":null},"finish_reason":"stop"}],"created":1755750392,"model":"Qwen/Qwen3-0.6B","object":"chat.completion","usage":{"prompt_tokens":10,"completion_tokens":131,"total_tokens":141}}
```

tzulingk commented on components/backends/vllm/src/dynamo/vllm/main.py:
Find out how duplicated registrations happen from the 2 paths of test codes below
 mocker = MockerProcess(request, endpoint, mocker_args_file)
 mocker.__enter__()
First Registration
ManagedProcess.__enter__()
-> ManagedProcess._start_process()
-> self.proc = subprocess.Popen(self.command, ...) // python -m dynamo.mocker --model-path ... --endpoint ...
-> worker() // components/backends/mocker/src/dynamo/mocker/main.py
-> run_input() // dynamo_llm::entrypoint::input::run_input()
-> endpoint::run()
-> endpoint.endpoint_builder().handler(ingress_chat).start(); // register metrics

Second Registration
ManagedProcess.__enter__()
-> ManagedProcess._start_process()
-> self.proc = subprocess.Popen(self.command, ...) // python -m dynamo.mocker --model-path ... --endpoint ...
-> worker() // components/backends/mocker/src/dynamo/mocker/main.py
-> engine_config = await make_engine(runtime, entrypoint_args) // bind to make_engine in lib/bindings/python/rust/llm/entrypoint.rs
-> select_engine()
-> dynamo_llm::mocker::engine::make_mocker_engine()
-> MockVllmEngine::new()
-> AnnotatedMockEngine::new()
-> MockVllmEngine::start()
-> MockVllmEngine::start_metrics_publishing()

let metrics_labels = Some(&[("model", "mock_model")][..]);
if let Err(e) = publisher
    .create_endpoint(comp.clone(), metrics_labels) // register metrics internally
    .await
So by removing the metrics_labels in MockVllmEngine, we can avoid having different labels set for the same metric.

nnshah1 commented on components/backends/vllm/src/dynamo/vllm/main.py:
```diff
@@ -178,7 +184,9 @@ async def init(runtime: DistributedRuntime, config: Config):
 
-    factory = StatLoggerFactory(component, config.engine_args.data_parallel_rank or 0)
+    factory = StatLoggerFactory(
+        component, config.engine_args.data_parallel_rank or 0, [("model", config.model)]
```
```suggestion
        component, config.engine_args.data_parallel_rank or 0, metrics_labels= [("model", config.model)]
```

nnshah1 commented on components/backends/vllm/src/dynamo/vllm/publisher.py:
```diff
@@ -129,15 +136,21 @@ def log_engine_initialized(self) -> None:
         if self.dp_rank != dp_rank:
             return NullStatLogger()
-        logger = DynamoStatLoggerPublisher(self.component, dp_rank)
+        logger = DynamoStatLoggerPublisher(self.component, dp_rank, self.labels)
```
```suggestion
        logger = DynamoStatLoggerPublisher(self.component, dp_rank, metrics_labels=self.metrics_labels)
```

nnshah1 commented on examples/multimodal/components/worker.py:
```diff
@@ -140,7 +140,9 @@ def setup_vllm_engine(self, component: Component, endpoint: Endpoint):
-            component, self.engine_args.data_parallel_rank or 0
+            component,
+            self.engine_args.data_parallel_rank or 0,
+            [("model", self.engine_args.model)],
```
```suggestion
            metrics_labels = [("model", self.engine_args.model)],
```

tzulingk committed:
Update examples/multimodal/components/worker.py

Co-authored-by: Neelay Shah <neelays@nvidia.com>
Signed-off-by: Tzu-Ling Kan <tzulingk@nvidia.com>

tzulingk committed:
Update components/backends/vllm/src/dynamo/vllm/main.py

Co-authored-by: Neelay Shah <neelays@nvidia.com>
Signed-off-by: Tzu-Ling Kan <tzulingk@nvidia.com>

tzulingk committed:
Update components/backends/vllm/src/dynamo/vllm/publisher.py

Co-authored-by: Neelay Shah <neelays@nvidia.com>
Signed-off-by: Tzu-Ling Kan <tzulingk@nvidia.com>

tzulingk committed:
Update components/backends/vllm/src/dynamo/vllm/publisher.py

Co-authored-by: Neelay Shah <neelays@nvidia.com>
Signed-off-by: Tzu-Ling Kan <tzulingk@nvidia.com>

tzulingk committed:
Update components/backends/vllm/src/dynamo/vllm/publisher.py

Co-authored-by: Neelay Shah <neelays@nvidia.com>
Signed-off-by: Tzu-Ling Kan <tzulingk@nvidia.com>

tzulingk committed:
Merge branch 'main' into tzulingk/model_label_dis_444

tzulingk committed:
Merge branch 'main' into tzulingk/model_label_dis_444

tzulingk committed:
Merge branch 'main' into tzulingk/model_label_dis_444

tzulingk committed:
fmt python files

tzulingk committed:
python fmt.

tzulingk committed:
pre-commit run --all-files results are different from my local black run, manuallly format it.

tzulingk committed:
Merge branch 'main' into tzulingk/model_label_dis_444
